File: node/tests/key-integration.test.ts
/**
 * Live integration tests for JACS key rotation and versioned fetch operations.
 *
 * Gated behind HAI_LIVE_TEST=1. Requires a running HAI API at
 * HAI_URL (defaults to http://localhost:3000).
 *
 * Run:
 *   HAI_LIVE_TEST=1 HAI_URL=http://localhost:3000 npx vitest run tests/key-integration.test.ts
 */

import { describe, it, expect, beforeAll } from 'vitest';
import { HaiClient } from '../src/index.js';
import { generateTestKeypair as generateKeypair } from './setup.js';

const LIVE = process.env.HAI_LIVE_TEST === '1';
const API_URL = process.env.HAI_URL || 'http://localhost:3000';

describe.skipIf(!LIVE)('Key integration (live API)', () => {
  let client: HaiClient;
  const agentName = `node-key-integ-${Date.now()}`;
  let jacsId: string;
  let agentId: string;

  // -------------------------------------------------------------------------
  // Setup: register agent
  // -------------------------------------------------------------------------

  beforeAll(async () => {
    const keypair = generateKeypair();
    client = await HaiClient.fromCredentials(agentName, keypair.privateKeyPem, {
      url: API_URL,
      privateKeyPassphrase: 'keygen-password',
    });

    const ownerEmail = process.env.HAI_OWNER_EMAIL || 'jonathan@hai.io';
    const result = await client.register({
      description: 'Node SDK key integration test agent',
      ownerEmail,
    });

    expect(result.success).toBe(true);
    jacsId = result.jacsId!;
    agentId = result.agentId!;
    console.log(`Registered agent: jacsId=${jacsId}, agentId=${agentId}`);
  }, 30_000);

  // -------------------------------------------------------------------------
  // Test: register then fetch key matches
  // -------------------------------------------------------------------------

  it('should fetch remote key matching registration', async () => {
    const key = await client.fetchRemoteKey(jacsId, 'latest');
    expect(key.jacsId || key.publicKey).toBeTruthy();
    expect(key.publicKey).toBeTruthy();
    expect(key.algorithm).toBeTruthy();
  });

  // -------------------------------------------------------------------------
  // Test: fetch key by hash
  // -------------------------------------------------------------------------

  it('should fetch key by hash', async () => {
    const key = await client.fetchRemoteKey(jacsId, 'latest');
    if (!key.publicKeyHash) {
      console.warn('Server did not return publicKeyHash, skipping');
      return;
    }

    const byHash = await client.fetchKeyByHash(key.publicKeyHash);
    expect(byHash.publicKey).toBe(key.publicKey);
    expect(byHash.algorithm).toBe(key.algorithm);
  });

  // -------------------------------------------------------------------------
  // Test: fetch key by email
  // -------------------------------------------------------------------------

  it('should fetch key by email after registration', async () => {
    // Username is now claimed during registration (one-step flow).
    const email = `${agentName}@hai.ai`;

    let byEmail;
    try {
      byEmail = await client.fetchKeyByEmail(email);
    } catch {
      console.warn('FetchKeyByEmail failed (agent may not have email in test env), skipping');
      return;
    }

    expect(byEmail.jacsId).toBeTruthy();
    expect(byEmail.publicKey).toBeTruthy();
  });

  // -------------------------------------------------------------------------
  // Test: fetch all keys returns history
  // -------------------------------------------------------------------------

  it('should return key history with at least one entry', async () => {
    const history = await client.fetchAllKeys(jacsId);
    expect(history.jacsId || history.total).toBeTruthy();
    expect(history.total).toBeGreaterThanOrEqual(1);
    expect(history.keys.length).toBeGreaterThanOrEqual(1);
    expect(history.keys[0].publicKey).toBeTruthy();
  });

  // -------------------------------------------------------------------------
  // Test: fetch key by domain returns 404 for fake domain
  // -------------------------------------------------------------------------

  it('should return 404 for nonexistent domain', async () => {
    await expect(
      client.fetchKeyByDomain('nonexistent-test-domain-12345.invalid'),
    ).rejects.toThrow();
  });
});

File: go/key_integration_test.go
// Live integration tests for JACS key rotation and versioned fetch operations.
//
// Gated behind HAI_LIVE_TEST=1. Requires a running HAI API at
// HAI_URL (defaults to http://localhost:3000).
//
// Run:
//
//	HAI_LIVE_TEST=1 HAI_URL=http://localhost:3000 go test -run TestKeyIntegration -v -count=1

package haiai

import (
	"context"
	"fmt"
	"os"
	"testing"
	"time"
)

func TestKeyIntegration(t *testing.T) {
	if os.Getenv("HAI_LIVE_TEST") != "1" {
		t.Skip("set HAI_LIVE_TEST=1 to run live API tests")
	}

	apiURL := os.Getenv("HAI_URL")
	if apiURL == "" {
		apiURL = "http://localhost:3000"
	}

	ctx := context.Background()
	agentName := fmt.Sprintf("go-key-integ-%d", time.Now().UnixMilli())

	// ── Setup: register a fresh JACS agent ────────────────────────────────
	ownerEmail := os.Getenv("HAI_OWNER_EMAIL")
	if ownerEmail == "" {
		ownerEmail = "jonathan@hai.io"
	}

	reg, err := RegisterNewAgentWithEndpoint(ctx, apiURL, agentName, &RegisterNewAgentOptions{
		Description: "Go key integration test agent",
		OwnerEmail:  ownerEmail,
		Quiet:       true,
	})
	if err != nil {
		t.Fatalf("RegisterNewAgentWithEndpoint: %v", err)
	}

	jacsID := reg.JacsID
	if jacsID == "" {
		jacsID = reg.AgentID
	}
	t.Logf("Registered agent: jacs_id=%s", jacsID)

	// Set the HAI_KEYS_BASE_URL to the live API so Client methods use it.
	t.Setenv("HAI_KEYS_BASE_URL", apiURL)

	// ── Test: register then fetch key matches ─────────────────────────────
	t.Run("RegisterThenFetchKeyMatches", func(t *testing.T) {
		// FetchRemoteKeyFromURL is deprecated. Use Client.FetchRemoteKey instead.
		// This test requires a client with credentials.
		if reg.PrivateKeyPath == "" {
			t.Skip("no private key path in registration result")
		}
		password, err := ResolvePrivateKeyPassword()
		if err != nil {
			t.Skipf("ResolvePrivateKeyPassword: %v", err)
		}
		privKey, err := LoadPrivateKey(reg.PrivateKeyPath, password)
		if err != nil {
			t.Skipf("LoadPrivateKey: %v", err)
		}
		cl, err := NewClient(
			WithEndpoint(apiURL),
			WithJACSID(jacsID),
			WithHaiAgentID(reg.AgentID),
			WithPrivateKey(privKey),
		)
		if err != nil {
			t.Skipf("could not build client: %v", err)
		}
		key, err := cl.FetchRemoteKey(ctx, jacsID, "latest")
		if err != nil {
			t.Fatalf("FetchRemoteKey: %v", err)
		}
		if len(key.PublicKey) == 0 {
			t.Fatal("expected non-empty public key")
		}
		if key.Algorithm == "" {
			t.Fatal("expected non-empty algorithm")
		}
		t.Logf("Fetched key: algorithm=%s, hash=%s", key.Algorithm, key.PublicKeyHash)
	})

	// ── Test: fetch key by hash via Client ────────────────────────────────
	t.Run("FetchKeyByHashMatches", func(t *testing.T) {
		// NOTE: FetchKeyByHashFromURL is deprecated. Use Client.FetchKeyByHash instead.
		// This test now requires a Client; skip if unavailable.
		t.Skip("FetchKeyByHashFromURL deprecated; use Client-based test instead")
	})

	// ── Test: fetch key by email via Client ──────────────────────────────
	t.Run("FetchKeyByEmailMatches", func(t *testing.T) {
		// Need a client with credentials to claim username.
		if reg.PrivateKeyPath == "" {
			t.Skip("no private key path in registration result")
		}

		// Load private key from path
		password, err := ResolvePrivateKeyPassword()
		if err != nil {
			t.Skipf("ResolvePrivateKeyPassword: %v", err)
		}
		privKey, err := LoadPrivateKey(reg.PrivateKeyPath, password)
		if err != nil {
			t.Skipf("LoadPrivateKey: %v", err)
		}

		// Build a client
		cl, err := NewClient(
			WithEndpoint(apiURL),
			WithJACSID(jacsID),
			WithHaiAgentID(reg.AgentID),
			WithPrivateKey(privKey),
		)
		if err != nil {
			t.Skipf("could not build client: %v", err)
		}

		claim, err := cl.ClaimUsername(ctx, reg.AgentID, agentName)
		if err != nil {
			t.Skipf("could not claim username: %v", err)
		}

		email := claim.Email
		if email == "" {
			t.Skip("no email returned from ClaimUsername")
		}

		// Use Client method (FFI-backed) instead of deprecated FetchKeyByEmailFromURL
		byEmail, err := cl.FetchKeyByEmail(ctx, email)
		if err != nil {
			t.Fatalf("FetchKeyByEmail: %v", err)
		}
		if len(byEmail.PublicKey) == 0 {
			t.Fatal("expected non-empty public key from email lookup")
		}
	})

	// ── Test: fetch all keys returns history ──────────────────────────────
	t.Run("FetchAllKeysReturnsHistory", func(t *testing.T) {
		// NOTE: FetchAllKeysFromURL is deprecated. Use Client.FetchAllKeys instead.
		// This test now requires a Client; skip if unavailable.
		t.Skip("FetchAllKeysFromURL deprecated; use Client-based test instead")
	})

	// ── Test: fetch key by domain returns 404 for fake domain ─────────────
	t.Run("FetchKeyByDomain404ForFakeDomain", func(t *testing.T) {
		// FetchKeyByDomainFromURL is deprecated -- now returns error directly.
		_, err := FetchKeyByDomainFromURL(ctx, nil, apiURL, "nonexistent-test-domain-12345.invalid")
		if err == nil {
			t.Fatal("expected error for deprecated function")
		}
		t.Logf("Got expected error: %v", err)
	})
}
